### PR TITLE
CI: sign and publish packages

### PR DESCRIPTION
**Pull request by cbix**
## Testing
Add repo to your `/etc/pacman.conf`:
```
[proaudio]
Server = https://arch.osamc.de/$repo/$arch
```
Fetch the current signing key (this may be subject to change soon!):
```com
$ wget https://arch.osamc.de/proaudio/osamc.gpg
```
Import and sign:
```com
# pacman-key --add osamc.gpg
# pacman-key --lsign-key 762AE5DB2B38786364BD81C4B9141BCC62D38EE5
```
You can now install packages from the repo using `pacman -Sy`.

## TODO
- [x] set packager id
- sign + publish
  - [x] sign packages using [drone-gpgsign](https://plugins.drone.io/drone-plugins/drone-gpgsign/) plugin
  - [x] `repo-add` all built packages
  - [x] scp to arch.osamc.de
  - [x] limit to `master` after testing
  - successful run log: https://ci.cbix.de/osam-cologne/archlinux-proaudio/138/1
- [x] use live repo (untrusted) during build steps (to support dependencies within our repo)
- [x] sign + publish aarch64 packages
  - successful run log: https://ci.cbix.de/osam-cologne/archlinux-proaudio/139/2
- [x] add step that uses the live repo to skip already published packages
  - this is also a good practice to enforce _always_ updating the pkgrel in order to trigger rebuilds ;)
- [x] figure out how to remove old packages
  - there is an `update-db.sh` script that scans the db for packages that are not included in the (git) repo and removes them
- [x] (before merging) limit secrets and CI steps to `master` (providing CI secrets to PRs is a security issue) + create fresh signing + ssh keys
  - signing key ID: `762AE5DB2B38786364BD81C4B9141BCC62D38EE5` ([download](https://arch.osamc.de/proaudio/osamc.gpg))
- [x] document necessary CI and server setup (adding secrets, creating server SSH key and adding it to `authorized_keys`, web server configuration (`.htaccess` files?).

## Notes
- create gpg key: `gpg --quick-generate-key 'OSAMC <https://github.com/osam-cologne/archlinux-proaudio> (trusted CI signing key)' ed25519 sign 5y`
- export secret: `gpg --export-secret-key $KEY_ID | base64 -w0`
- export public: `gpg --export --armor $KEY_ID > osamc.gpg`
- create ssh keypair: `ssh-keygen -t ed25519 -C ci@osamc -f ./id_osamc -N ""`

### CODE REVIEW

**Comment by SpotlightKid** [comment]
Should we create an arch.osamc.de master key to sign member keys and the key used by the CI?

**Comment by cbix** [comment]
> Should we create an arch.osamc.de master key to sign member keys and the key used by the CI?

@SpotlightKid yes! Good idea 👍 Haven't really wrapped my head around how in the end users will import this key easily in order to use the repo...

My approach at creating the repo db was actually to always create a fresh db and add all built packages when pushing to `master`, but that would mean we can't just manually tweak that db and add packages without using the CI. What do you think?

**Comment by SpotlightKid** [comment]
> Haven't really wrapped my head around how in the end users will import this key easily in order to use the repo...

Like it is described here: https://wiki.archlinux.org/title/Pacman/Package_signing#Adding_unofficial_keys

We can copy the relevant parts into our top-level readme or put it into a GH wiki page.

Then list we list our repo [here](https://wiki.archlinux.org/title/Unofficial_user_repositories) and insert a link in the description to the above instructions.

**Comment by SpotlightKid** [comment]
> [... ]always create a fresh db and add all built packages when pushing to `master`, [...]. What do you think?

Can't we cache the DB in the CI environment somewhere? If not, I would just download it with rsync from the repo server before doing the `repo-add`. If we sign the DB and check the sig after download that should make sure it stays consistent.

**Comment by cbix** [comment]
> [...] I would just download it with rsync from the repo server before doing the `repo-add`.

just added that to the TODO 👍

**Comment by cbix** [comment]
@SpotlightKid I hope my documentation matches your request :D

File: tools/README.md [hunk 1]
```diff
@@ -0,0 +1,66 @@
+# Behind the scenes
+
+## Continuous Integration
+
+Most of the tools in this directory are being used by the CI Build / Check / Publish pipeline, so here's a rough
```
Comment (SpotlightKid) [inline tools/README.md:5]: suggestion: "[...] pipeline, which uses [Drone.io](https://www.drone.io/], [...]"

**Review by SpotlightKid — CHANGES_REQUESTED** [review]

File: tools/README.md [hunk 2]
```diff
@@ -0,0 +1,66 @@
...
+and a private key stored as a secret in the CI system which is only available on pushes to `master`.
+The key is generated like this
+```
+$ gpg --quick-generate-key 'OSAMC <https://github.com/osam-cologne/archlinux-proaudio>' ed25519 sign 5y
+```
+Then the secret key is exported
+```
+gpg --export-secret-key $KEY_ID | base64 -w0
+```
+and stored as a Drone secret. The public key is exported using
```
[the remark this answers is not in the document]
Comment (cbix) [reply tools/README.md]: b837dea09fc2d2c3a92d2bf62550fdb7f22b9227

File: tools/README.md [hunk 3]
```diff
@@ -0,0 +1,66 @@
...
+### Update database
+This is basically using the stock `repo-add` and `repo-remove` which are shipped with Pacman.
+
+### Publish
+The [Drone SCP Plugin](https://plugins.drone.io/appleboy/drone-scp/) is used to push the resulting files to
+the server. This needs a private SSH key stored as another Drone secret:
+```
+ssh-keygen -t ed25519 -C ci@osamc -f ./id_osamc -N ""
+```
+The server is a simple web server and only requires all published files to sit in the same directory.
```
[the remark this answers is not in the document]
Comment (cbix) [reply tools/README.md]: b837dea09fc2d2c3a92d2bf62550fdb7f22b9227

File: tools/README.md (hunk 1 above)
Comment (cbix) [reply tools/README.md]: b837dea09fc2d2c3a92d2bf62550fdb7f22b9227

File: tools/README.md (hunk 2 above)
[the remark this answers is not in the document]
Comment (cbix) [reply tools/README.md]: (don't worry, I used `head -c 500 /dev/urandom | base64 -w0` to generate the dummy base64 "private key" 😜)

File: tools/README.md [hunk 4]
```diff
@@ -0,0 +1,91 @@
...
+$ drone secret add --name pkgsignkey --data '...5LAqcQp6zTA5lIc2tQpVf3F8+c=' osam-cologne/archlinux-proaudio
+```
+If you added a passphrase in the gpg step, also create a secret for that (`pkgsignkey-pass`).
+
+The public key is exported using
+```
+gpg --export --armor $KEY_ID > osamc.gpg
+```
+and can be [downloaded from the repo server](https://arch.osamc.de/proaudio/osamc.gpg)
+```
```
[the remark this answers is not in the document]
Comment (cbix) [reply tools/README.md:66]: Oops :D Will open another small PR including this fix later today!

**Pull request merged**